common.go

```diff
@@ -1,71 +1,171 @@
+// to find out the most probably language to return.
+func GetLanguage(filename string, content []byte) string {
+	candidates := make([]string, 0, 5)
+	for _, strategy := range strategies {
+		languages := strategy(filename, content)
+		if len(languages) == 1 {
+			return languages[0]
+		}
+
+		if len(languages) > 0 {
+			candidates = append(candidates, languages...)
+		}
 	}
-	auxiliaryLanguages = map[string]bool{
-		"Other": true, "XML": true, "YAML": true, "TOML": true, "INI": true,
-		"JSON": true, "TeX": true, "Public Key": true, "AsciiDoc": true,
-		"AGS Script": true, "VimL": true, "Diff": true, "CMake": true, "fish": true,
-		"Awk": true, "Graphviz (DOT)": true, "Markdown": true, "desktop": true,
-		"XSLT": true, "SQL": true, "RMarkdown": true, "IRC log": true,
-		"reStructuredText": true, "Twig": true, "CSS": true, "Batchfile": true,
-		"Text": true, "HTML+ERB": true, "HTML": true, "Gettext Catalog": true,
-		"Smarty": true, "Raw token data": true,
+
+	if len(candidates) == 0 {
+		return OtherLanguage
 	}
-)
 
-func init() {
-	for l, _ := range ignoredExtensions {
-		languagesByExtension[l] = []string{OtherLanguage}
+	lang := GetLanguageByClassifier(content, candidates, nil)
+	return lang
+}
+
+// GetLanguageByModeline returns the language of the given content looking for the modeline,
+// and safe to indicate the sureness of returned language.
+func GetLanguageByModeline(content []byte) (lang string, safe bool) {
+	languages := getLanguagesByModeline("", content)
+	if len(languages) == 0 {
+		lang = OtherLanguage
+		return
 	}
 
-	ExtensionsByLanguage = reverseStringListMap(languagesByExtension)
+	return languages[0], true
 }
 
-// GetLanguageExtensions returns the different extensions being used by the
-// language.
-func GetLanguageExtensions(language string) []string {
-	return ExtensionsByLanguage[language]
+// GetLanguageByFilename returns a language based on the given filename, and safe to indicate
+// the sureness of returned language.
+func GetLanguageByFilename(filename string) (lang string, safe bool) {
+	languages := getLanguagesByFilename(filename, nil)
+	if len(languages) == 0 {
+		lang = OtherLanguage
+		return
+	}
+
+	return languages[0], true
 }
 
-// GetLanguage return the Language for a given filename and file content.
-func GetLanguage(filename string, content []byte) string {
-	if lang, safe := GetLanguageByModeline(content); safe {
-		return lang
+func getLanguagesByFilename(filename string, content []byte) []string {
+	return languagesByFilename[filename]
+}
+
+// GetLanguageByShebang returns the language of the given content looking for the shebang line,
+// and safe to indicate the sureness of returned language.
+func GetLanguageByShebang(content []byte) (lang string, safe bool) {
+	languages := getLanguagesByShebang("", content)
+	if len(languages) == 0 {
+		lang = OtherLanguage
+		return
 	}
 
-	if lang, safe := GetLanguageByFilename(filename); safe {
-		return lang
+	lang = languages[0]
```

Comment:
This code is equivalent to the code in `GetLanguageByFilename`, right? If so, be consistent with the style.
It may be even better to extract the pattern of shared code to a private function, to avoid the unnecessary duplication.